File: src/commonMain/resources/webapp/js/gameEngine.js
class NetworkGameEngine {

    constructor(color, mapSelected = "free") {
        this.color = color.substring(1);
        this.map = mapSelected;
    }

    startGame(roomId = "") {
        let url = "ws://" + window.location.host + "/games/snake/game";
        if (roomId !== "")
            url += "/" + roomId;
        url += "?colorId=" + this.color + "&mapName=" + this.map;
        this.socket = new WebSocket(url);

        this.socket.onopen = () => (console.log("Connection is set!"));
        this.socket.onclose = (event) => {
            if (event.wasClean) {
                console.log("Clean closed");
            } else {
                console.log("Connection reset");
            }
            console.log("Error code: " + event.code + ", reason: " + event.reason);
        }
    
        this.socket.onmessage = (event) => {
            let roomId = Number.parseInt(event.data);
            if  (!isNaN(roomId)) {
                $('.temp').html('Room ID <br>' + roomId);
                return;
            }

            let response = JSON.parse(event.data);
            let points = response.points;
            let ctx = getCanvasContext();
            clearCanvas(ctx);
            for (let point of points) {
                paint(point.x, point.y, '#' + point.color, ctx);
            }
        }
    
        this.socket.onerror = (event) => {
            console.log("Error: " + event.message);
        }
    }

    stopGame() {
        this.socket.close();
        clearCanvas(getCanvasContext());
    }

    sendData(direction) {
        let request = {
            "direction": direction
        }
        request = JSON.stringify(request);
        this.socket.send(request);
    }

}

class SingleGameEngine {

    constructor(color, map) {
        this.food = new Food();
        this.snake = new Snake(color);
        this.map = map;
    }

    startGame() {
        this.timer = setInterval(() => { this.step(); }, 10);
    }

    step() {
        let points = [];
        if (this.snake.intersects(this.food.cords)) {
            this.food.newCords(this.map.edges);
            this.snake.grow();
            if (this.snake.props.speed > 10 && this.snake.getSize() % 10 == 0) {
                this.snake.props.speed -= 10;
            }
        }

        if (this.snake.intersectsItself())
            this.snake.reborn();

        if (this.map.isIntersectedBy(this.snake))
            this.snake.reborn();

        this.snake.updateDirection();
        this.snake.updateCords();

        for (let element of this.snake.props.snake) {
            let {x, y} = element.getCords();
            points.push({ x: x, y: y, color: this.snake.props.color })
        }

        let edges = this.map.edges;
        for (let point of edges) {
            points.push({
                x: point.x,
                y: point.y,
                color: this.map.color
            })
        }

        points.push({x: this.food.cords.x, y: this.food.cords.y, color: this.food.color});

        let ctx = getCanvasContext();
        clearCanvas(ctx);
        for (let point of points) {
            paint(point.x, point.y, point.color, ctx);
        }
        
    }

    stopGame() {
        clearInterval(this.timer);
        clearCanvas(getCanvasContext());
    }

    sendData(direction) {
        let dir;
        switch(direction) {
            case "UP": dir = DIRECTION_UP; break;
            case "DOWN": dir = DIRECTION_DOWN; break;
            case "RIGHT": dir = DIRECTION_RIGHT; break;
            case "LEFT": dir = DIRECTION_LEFT; break;
        }
        this.snake.addDirection(dir);
    }

}

File: src/jsMain/kotlin/GameEngine.kt
import com.example.gameModes.GameMode
import com.example.maps.GameMap
import com.example.model.Direction
import com.example.model.Point
import com.example.model.Snake
import com.example.model.food.FoodWrapper
import com.example.utils.*
import kotlinx.browser.window
import kotlinx.serialization.json.Json
import org.w3c.dom.CloseEvent
import org.w3c.dom.ErrorEvent
import org.w3c.dom.WebSocket
import kotlin.js.Promise

interface GameEngine {
    fun startGame(): Promise<String?>
    fun stopGame()
    fun changeDirection(direction: String)
}

class SingleGameEngine(color: String, private val map: GameMap, private val gameMode: GameMode) : GameEngine {

    private var foods = mutableListOf(gameMode.spawnFood(map.edges))
    private val snake = Snake(color.substring(1), DEFAULT_SPEED)
    private var timer: Int? = null

    constructor(color: String, map: String, gameMode: String) : this(color, defineMap(map), defineGameMode(gameMode))

    init {
        gameMode.map = map
    }

    override fun startGame(): Promise<String?> {
       timer = window.setInterval(::step, 10)
        return Promise{ resolve, _ ->
            resolve(null)
        }
    }

    private fun step() {
        val points = ArrayList<Point>()
        if (snake.intersects(foods[0].cords)) {
            snake.eat(foods[0])
            foods[0] = gameMode.spawnFood(map.edges)
        }

        if (snake.intersectsItself())
            snake.reborn()

        if (map.isIntersected(snake))
            snake.reborn()

        snake.updateDirection()
        snake.updateCords()

        gameMode.apply(listOf(snake), foods)

        for (element in snake.snake) {
            val cords = element.getCords()
            points.add(Point(cords, snake.color))
        }

        for (point in map.edges) {
            points.add(Point(point, map.color))
        }

        val ctx = getCanvasContext()
        clearCanvas(ctx)
        for (point in points) {
            paint(point.x, point.y, "#${point.color}", ctx)
        }
        paintFood(FoodWrapper(foods[0]), ctx)
    }

    override fun stopGame() {
        if (timer != null) {
            window.clearInterval(timer!!)
            clearCanvas(getCanvasContext())
        }
    }

    override fun changeDirection(direction: String) {
        val dir = when (direction) {
            "UP" -> Direction.UP
            "DOWN" -> Direction.DOWN
            "RIGHT" -> Direction.RIGHT
            "LEFT" -> Direction.LEFT
            else -> return
        }
        snake.addDirection(dir)
    }

}

class NetworkGameEngine(color: String, private val mapSelected: String = "free",
                        private val roomId: String = "", private val gameMode: String) : GameEngine {

    private val color = color.substring(1)
    private var socket: WebSocket? = null

    override fun startGame(): Promise<String> {
        var path = "ws://${window.location.host}/games/snake/game"
        if (roomId != "") {
            path += "/$roomId"
        }
        path += "?colorId=$color&mapName=$mapSelected&gameMode=$gameMode"

        return Promise{ resolve, reject ->
            this.socket = WebSocket(path)

            this.socket!!.onopen = {
                console.log("Connection is set!")
            }
            this.socket!!.onclose = { e ->
                val event = e as CloseEvent
                if (event.wasClean) {
                    console.log("Clean closed")
                } else {
                    console.log("Connection reset")
                    reject(Throwable("Troubles with connection"))
                }
                console.log("Error code: " + event.code + ", reason: " + event.reason)
            }

            this.socket!!.onmessage = { event ->
                val receivedText = event.data.toString()
                try {
                    receivedText.toInt()
                    resolve(receivedText)
                } catch (e: Exception) {
                    val response = Json.decodeFromString(Response.serializer(), receivedText)
                    val points = response.points
                    val ctx = getCanvasContext()
                    clearCanvas(ctx)
                    for (point in points) {
                        paint(point.x, point.y, "#${point.color}", ctx)
                    }
                    val food = response.foodWrapper
                    paintFood(food, ctx)
                }
            }

            this.socket!!.onerror = { e ->
                val event = e as ErrorEvent
                console.log("Error" + event.message)
                reject(Throwable("Error happened. ${e.message}"))
            }
        }
    }

    override fun stopGame() {
        socket!!.close()
        console.log("socket closed")
        clearCanvas(getCanvasContext())
    }

    override fun changeDirection(direction: String) {
        val dir = defineDirection(direction) ?: return
        val request = Json.encodeToString(Request.serializer(), Request(dir))
        socket!!.send(request)
    }
}